src/core/server/mocks.ts

```diff
@@ -86,6 +86,8 @@ function pluginInitializerContextMock<T>(config: T = {} as T) {
   return mock;
 }
 
+type CoreSetupMockType = MockedKeys<CoreSetup> & jest.Mocked<Pick<CoreSetup, 'getStartServices'>>;
+
```

Comment:
I had a hard time adding `getStartServices` to the `CoreSetup` mock. `MockedKeys` does not properly handles functions and resulted in the mock not being compatible with the concrete implementation. Using `jest.Mocked` or `DeeplyMockedKeys` both resulted on errors on tests that were overriding read-only properties (that seems to be preserved by both but not by `MockedKeys`), such as
https://github.com/elastic/kibana/blob/56041f03ada9730e701508dbc10032b69de9f218/x-pack/plugins/licensing/server/plugin.test.ts#L54-L55

So I had to compose this custom type instead.

Reply:
It sounds like something we will face in the future again when marking more API points as `readonly`. We need to define a common pattern for configuring mocks at any deep level. Re-writing a property is the most obvious way, but create this incompatibility. Options:
- a mock factory accepts an object configuring mocks at any deep `coreMock.createSetup({ elasticsearch: { dataClient: ... } });` That would quick grow with the number of nested objects.
- provide configurable mock factory (as a builder pattern, for example) `coreMock.createSetup().withElasticsearchDataClient(...).withSomethingelse().build()`

Reply:
After a quick look, this is not the first time this occurs (the exemple is actually the root cause if this specific issue)

https://github.com/elastic/kibana/blob/2d62ff2cbfb0bed6f5658e8dd39ef1207273ca46/src/core/server/elasticsearch/elasticsearch_service.mock.ts#L46-L51

We are also inconsistent in the types we actually return from mocks. Some are explicitly mock instance, some are not

https://github.com/elastic/kibana/blob/ea9a7b8a16abbee97ddf39f91e4cf3c3b7253b42/src/core/server/mocks.ts#L121-L122

vs

https://github.com/elastic/kibana/blob/ea9a7b8a16abbee97ddf39f91e4cf3c3b7253b42/src/core/server/mocks.ts#L131-L132

> That would quick grow with the number of nested objects.

It does. however the (manually coded) factory pattern present the same (if not bigger) problem in term of maintenance. I'm not the biggest TS expert, but can't we have an utility type that would remove readonly from the mock definition while preserving compatibility with the concrete type?

Reply:
Hum, I was trying things and was not understanding why I couldnt override `core.elasticsearch.dataClient` from inside core... And I just saw that the `MockedKeys` type is not the same between `src` and `x-pack`... This explain why this is possible in the licensing plugin tests:

https://github.com/elastic/kibana/blob/eb03fd8c5f3e4f79b503dcccbc4589812e036f10/typings/index.d.ts#L32

https://github.com/elastic/kibana/blob/643f15c5d975eca1df4be948c01a78a749101296/x-pack/typings/index.d.ts#L30

Reply:
This is outside of the scope of this PR. Created https://github.com/elastic/kibana/issues/55189 to continue the discussion